### ISSUE TITLE
case tags: re-specify structure, simplify (str/str mapping?)

### ISSUE THREAD

**Issue by jgehrcke** [conbench/conbench#940]
Today: database allows any JSON document as value: https://github.com/conbench/conbench/blob/9237373f99b2aa04906496e0393eabc0ec1ecd45/conbench/entities/case.py#L12

JSON schema validation also kind of allows everything (not sure what marshmallow.fields.Dict() really allows, but probably nesting): https://github.com/conbench/conbench/blob/9237373f99b2aa04906496e0393eabc0ec1ecd45/conbench/entities/benchmark_result.py#L580

Nesting might hurt us. None-values might raise complications. Mixed value types increase complexity for which value?

I think we should make this a str/str mapping. Empty string values? Not allowed. Then we know exactly what a key/value pair looks like.

Looks like @jonkeane agrees: https://github.com/conbench/conbench/pull/937#discussion_r1145196482

**Comment by jonkeane** [comment]
I'm fine with this — I don't know of any examples of someone using something nested right now and better to cut off that possibility (+ needing to deal with it) sooner than later.

> Empty string values? Not allowed. Then we know exactly what a key/value pair looks like.

When you say empty string here are you thinking `""` or `null`s?

**Comment by jgehrcke** [comment]
> When you say empty string here are you thinking "" or nulls?

When I wrote the above I meant to not allow JSON null as well as to not allow empty string "".

I thought _only_ not including a key should semantically mean "not set".

The special value `""` (empty string) often also means "not set", and if we allow that then we have more than one way to represent "not set".

In our discussion I think I just convinced myself that it's ok/good to allow for empty string, but now I am not so convinced anymore. I will give this more thought during texting/implementing.

The one big/important design choice is: that absence of a key name still allows for that labelset to be part of the set of all case permutations, as long as the benchmark _name_ is in common.
